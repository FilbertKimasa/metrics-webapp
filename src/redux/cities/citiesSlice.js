import { v4 as uuidv4 } from 'uuid';
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const airApiKey = 'dea85b82ceccf415bffc6e9e3c2ca0fa';
const coordinateApiKey = 'rB16lee6ZvybB0ZCrsrzwQ==6odUeEwciclckq71';

const initialState = {
  cities: [
    'Cairo',
    'Lagos',
    'Kinshasa',
    'Johannesburg',
    'Nairobi',
    'Casablanca',
    'Addis Ababa',
    'Abidjan',
    'Alexandria',
    'Khartoum',
    'Durban',
    'Accra',
    'Dar es Salaam',
    'Algiers',
    'Cape Town',
    'Luanda',
    'Pretoria',
    'Maputo',
    'Dakar',
    'Harare',
  ],
  citiesData: [],
};

const cityAirData = async (lat, long) => {
  const response = await fetch(
    `https://api.openweathermap.org/data/2.5/air_pollution?lat=${lat}&lon=${long}&appid=${airApiKey}`,
  ).then((resp) => resp.json());
  return response;
};

const cityCoordinates = async (city) => {
  const response = await fetch(
    `https://api.openweathermap.org/geo/1.0/direct?q=${city}&limit=1&appid=${coordinateApiKey}`,
  ).then((resp) => resp.json());
  return response;
};

const GET_DATA = 'airquality/getindex';
const getCitiesData = createAsyncThunk(GET_DATA, async (_, thunkAPI) => {
  const citiesDataArray = [];
  const { cities } = thunkAPI.getState();
  cities.forEach((city) => {
    citiesDataArray.push(
      cityCoordinates(city).then(async (cityCoord) => {
        const { lat } = cityCoord[0];
        const long = cityCoord[0].lon;
        const cityData = await cityAirData(lat, long);
        return { id: uuidv4(), city, data: cityData.list[0] };
      }),
    );
  });
  const citiesData = await Promise.all(citiesDataArray).then(
    (result) => result,
  );
  return citiesData;
});

const citiesSlice = createSlice({
  name: 'handleIndexes',
  initialState,
  extraReducers: (builder) => {
    builder.addCase(getCitiesData.fulfilled, (state, action) => {
      state.cities.citiesData = action.payload;
    });
  },
});

export { getCitiesData };
export default citiesSlice.reducer;
