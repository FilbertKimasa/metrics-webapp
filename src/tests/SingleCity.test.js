/* eslint-disable import/order */
import { getCitiesData, cityAirData } from '../redux/cities/citiesSlice';
import axios from 'axios';

jest.mock('axios');

describe('getCitiesData', () => {
  it('should fetch city coordinates and air quality data for each city', async () => {
    const mockDispatch = jest.fn();
    const mockGetState = jest.fn(() => ({ cities: { cities: ['Cairo'] } }));
    const mockCityCoordinates = jest
      .fn()
      .mockResolvedValue([{ lat: 30, lon: 31 }]);
    const mockCityAirData = jest.fn().mockResolvedValue({ airQuality: 'Good' });
    axios.get = jest
      .fn()
      .mockResolvedValueOnce({ data: [{ lat: 30, lon: 31 }] })
      .mockResolvedValueOnce({ data: { airQuality: 'Good' } });

    await getCitiesData(undefined, {
      dispatch: mockDispatch,
      getState: mockGetState,
    });

    expect(mockCityCoordinates).toHaveBeenCalledWith('Cairo');
    expect(mockCityAirData).toHaveBeenCalledWith(30, 31);
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'airquality/getindex/fulfilled',
      payload: [
        {
          id: expect.any(String),
          city: 'Cairo',
          data: { airQuality: 'Good' },
        },
      ],
    });
  });

  it('should handle errors when fetching city coordinates', async () => {
    const mockDispatch = jest.fn();
    const mockGetState = jest.fn(() => ({ cities: { cities: ['Cairo'] } }));
    const mockCityCoordinates = jest.fn().mockRejectedValue('failed to fetch');
    axios.get = jest.fn().mockRejectedValue('failed to fetch');

    await getCitiesData(undefined, {
      dispatch: mockDispatch,
      getState: mockGetState,
    });

    expect(mockCityCoordinates).toHaveBeenCalledWith('Cairo');
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'airquality/getindex/fulfilled',
      payload: [null],
    });
  });
});

describe('cityAirData', () => {
  it('should fetch air quality data for a given latitude and longitude', async () => {
    const lat = 30;
    const lon = 31;
    const mockResponse = { airQuality: 'Good' };
    axios.get.mockResolvedValueOnce({ data: mockResponse });

    const result = await cityAirData(lat, lon);

    expect(result).toEqual(mockResponse);
    expect(axios.get).toHaveBeenCalledWith(
      `https://api.openweathermap.org/data/2.5/air_pollution?lat=${lat}&lon=${lon}&appid=862090a665075dd09b646a7cca4e4e1e`,
    );
  });
});
