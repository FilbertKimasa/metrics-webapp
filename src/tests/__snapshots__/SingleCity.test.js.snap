// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<div
  className="display-grid"
>
  <img
    alt="map"
    className="map"
    src="africa-map.png"
  />
  <p
    className="heading"
  >
    Air Quality In Some African Big Cities
  </p>
</div>
`;
